### ISSUE TITLE
Redesigned Training Plan: Semantic Similarity for Morpheme Embeddings

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#58]
# Redesigned Training Plan: Semantic Similarity for Morpheme Embeddings

## Why Previous Approach Failed (Issue #57)

**MLM (Masked Language Modeling) is wrong for Esperanto**:
- MLM trains models to predict grammar/form from context
- Esperanto grammar is 100% deterministic (16 rules, no exceptions)
- Training loss INCREASED each epoch (6.96 → 7.24 → 7.28)
- The model was trying to learn what it can't predict (deterministic rules)

**Solution**: Train for **semantic similarity**, not grammatical prediction.

---

### Component Overview

| Component | Dimension | Training Method | Why |
|-----------|-----------|-----------------|-----|
| **Root embeddings** | 128d | Contrastive learning | Learn MEANING from semantic context |
| **Semantic prefixes** (8) | 128d | Composition learning | Transform root meanings |
| **Semantic suffixes** (15) | 128d | Composition learning | Derive new concepts |
| **Grammatical prefixes** (4) | 16d | **NO TRAINING** | Deterministic features |
| **Aspectual suffixes** (13) | 16d | **NO TRAINING** | Deterministic features |
| **Endings** (17) | 16d | **NO TRAINING** | Deterministic features |

**Parameter Count**:
- Roots: 5,000 × 128d = 640K
- Semantic prefixes: 8 × 128d = 1K
- Semantic suffixes: 15 × 128d = 2K
- Grammatical (frozen): 34 × 16d = 0.5K
- **Total: ~650K learned parameters** (vs 1.29M before)

---

### Objective 1: Root Semantic Similarity (Primary)

**Goal**: Roots that appear in similar contexts should have similar embeddings.

**Training Data Generation**:
```
Positive pairs: Sentences from SAME Wikipedia article (topic similarity)
Negative pairs: Sentences from DIFFERENT Wikipedia articles (random)
```

**Loss Function**: Cosine Similarity Loss (existing from `train_semantic_similarity.py`)
```python
# Normalize embeddings
emb_a = F.normalize(emb_a, dim=-1)
emb_b = F.normalize(emb_b, dim=-1)

# Predicted similarity
pred_sim = torch.sum(emb_a * emb_b, dim=-1)

# MSE loss between predicted and target
loss = F.mse_loss(pred_sim, target_sim)
```

**Key Insight**: We're not predicting the next word - we're learning that sentences with similar roots should cluster together.

---

### Objective 2: Semantic Affix Composition (Secondary)

**Goal**: Learn how semantic affixes transform root meanings.

**Training Data Generation**:
```
For each affix application in corpus:
  - "hundejo" (dog-place) should be near words about places
  - "malbona" (un-good) should be opposite of "bona"
  - "dometo" (house-DIM) should be near "domo" but smaller semantically
```

**Loss Function**: Triplet Loss with margin
```python
# anchor = "hundejo" embedding
# positive = "lernejo", "vendejo" (other -ej words, semantically similar)
# negative = "hundo" (without suffix, different meaning category)

loss = max(0, margin + sim(anchor, negative) - sim(anchor, positive))
```

**Training Strategy**:
- Freeze root embeddings initially (trained in Objective 1)
- Train only semantic affix embeddings
- Use affix groupings: all -ej words vs all -ul words vs all -et words

---

### Objective 3: Grammatical Features (NO TRAINING)

**These are NOT learned - they're deterministic feature vectors**:

```python
class GrammaticalFeatures(nn.Module):
    def __init__(self):
        # Fixed, not trained
        self.register_buffer('ending_features', self._create_ending_features())
        self.register_buffer('aspect_features', self._create_aspect_features())
        
    def _create_ending_features(self):
        # One-hot or hand-crafted features
        # -o: [1,0,0,0, 0,0, 0,0,0,0,0,0] = noun, nominative, singular
        # -ojn: [1,0,0,0, 1,0, 1,0,0,0,0,0] = noun, accusative, plural
        return torch.tensor([...])  # 17 endings × 16 features
```

**Reasoning**: Esperanto grammar is perfectly regular. Training a model to predict "-as" vs "-is" is wasted capacity - the parser already knows this deterministically.

---

### Phase 1: Root Embedding Training

1. **Generate training pairs** from Wikipedia corpus:
   - Same article = positive (similarity 0.8-1.0)
   - Different article = negative (similarity 0.0-0.3)
   - Random sampling with ratio 1:5 positive:negative
   
2. **Extract root sequences** from ASTs (not full words)
   
3. **Train with CosineSimilarityLoss** for 10-20 epochs
   
4. **Evaluate**: 
   - Nearest neighbor quality (hund → kat, hund → best, hund → etc.)
   - Semantic clustering (animals together, verbs together)

### Phase 2: Semantic Affix Training

1. **Freeze root embeddings** from Phase 1

2. **Extract affix application patterns** from corpus:
   ```
   "hundejo" → {root: hund, suffix: ej, ending: o}
   "lernejo" → {root: lern, suffix: ej, ending: o}
   ```

3. **Train with TripletLoss**:
   - Anchor: word with affix
   - Positive: other words with SAME affix type
   - Negative: base root or different affix type

4. **Evaluate**:
   - Affix cluster separation (all -ej words cluster vs all -ul words)
   - Composition quality: embed("hund") + embed("-ej") ≈ embed("hundejo")

### Phase 3: Integration & Eval

1. **Combine all components** into MorphemeAwareEmbedding
2. **Freeze grammatical features** (never trained)
3. **End-to-end evaluation** on retrieval task
4. **Compare to baseline** (FastText embeddings)

---

## Implementation Tasks

- [ ] Create `scripts/generate_semantic_pairs.py` - Generate positive/negative pairs from Wikipedia articles
- [ ] Modify `scripts/train_semantic_similarity.py` to use root-only embeddings
- [ ] Create `scripts/train_affix_composition.py` - Phase 2 affix training
- [ ] Add `GrammaticalFeatures` class with frozen deterministic features
- [ ] Create evaluation scripts for semantic quality
- [ ] Integrate into retrieval pipeline

---

## Success Criteria

1. **Training loss DECREASES** (not increases like MLM)
2. **Semantic nearest neighbors make sense** (hund near kat, not near libr)
3. **Affix composition works**: root + affix ≈ derived word
4. **Retrieval quality improves** over baseline FastText

---

## Files Involved

- `klareco/embeddings/morpheme_aware.py` - Existing architecture (good)
- `config/affix_classification.json` - Affix categories (good)
- `scripts/train_semantic_similarity.py` - Template for training loop
- `scripts/generate_training_pairs.py` - Template for pair generation

## Related Issues

- #57 (Wrong training objective - CLOSED as superseded by this)
- #51-56 (Closed - superseded by this redesign)

**Comment by marctjones** [comment]
Superseded by #65 (Epic: Fundamento-Centered Training). The new plan uses cross-lingual semantic anchoring from the Fundamento instead of the semantic similarity approach.
